docs/source/cuda/external-memory.rst

```diff
@@ -0,0 +1,280 @@
+.. _cuda-emm-plugin:
+
+=================================================
+External Memory Management (EMM) Plugin interface
+=================================================
+
+The :ref:`CUDA Array Interface <cuda-array-interface>` enables sharing of data
+between different Python libraries that access CUDA devices. However, each
+library manages its own memory distinctly from the others. For example:
+
+- By default, Numba allocates memory on CUDA devices by interacting with the
+  CUDA driver API to call functions such as ``cuMemAlloc`` and ``cuMemFree``,
+  which is suitable for many use cases.
+- The RAPIDS libraries (cuDF, cuML, etc.) use the `RAPIDS Memory Manager (RMM)
+  <https://github.com/rapidsai/rmm>`_ for allocating device memory.
+- `CuPy <https://cupy.chainer.org/>`_ includes a `memory pool implementation
+  <https://docs-cupy.chainer.org/en/stable/reference/memory.html>`_ for both
+  device and pinned memory.
+
+When multiple CUDA-aware libraries are used together, it may be preferable for
+Numba to defer to another library for memory management. The EMM Plugin
+interface facilitates this, by enabling Numba to use another CUDA-aware library
+for all allocations and deallocations.
+
+An EMM Plugin is used to facilitate the use of an external library for memory
+management. An EMM Plugin can be a part of an external library, or could be
+implemented as a separate library.
+
+
+Overview of External Memory Management
+======================================
+
+When an EMM Plugin is in use (see :ref:`setting-emm-plugin`), Numba will make
+allocation and deallocation through the Plugin. It will never directly call
+functions such as ``cuMemAlloc``, ``cuMemFree``, etc.
```

Comment:
Just to clarify this scenario, is it:

- An extension library uses `numba.cuda` to call `cuMemAlloc` to allocate memory, and then later calls `cuMemFree` on its own allocations,
- An extension library uses `numba.cuda` to call `cuMemFree` on memory pointers provided by the EMM Plugin, or
- Something else?

In any case, the first scenario is fine - behaviour will be well defined, but if a third-party library doesn't use the same allocator that Numba does, there's not a lot that can be done to stop it. The second is obviously wrong, since the pointers might well be part of a pool, and the EMM plugin will be expecting to free them at some point, but I'm finding it hard picture a use case where you would expect to do this.

Reply:
I think this is ok on second thoughts. Re-read it as "Numba internally will make allocations through the EMM" and that's fine, further nothing precludes a users using `numba.cuda.cuMemAlloc/cuMemFree` independently.